_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,133 @@
+[Solidity version 0.8.17](https://github.com/ethereum/solidity/releases/tag/v0.8.17), released on September 07, 2022, provides a fix.
+
+We assigned the bug a severity of "medium/high".
+
+
+## Which Contracts are Affected?
+
+A prerequisite for a contract to be affected is the use of ``return(...)`` or ``stop()`` in an inline assembly block (the Solidity ``return`` statement is not affected). Such an inline assembly call will not return (as in continue control flow after the statement), but instead result in an early successful (i.e. non-reverting) termination of the entire transaction.
+
+If this termination of the transaction happens conditionally within a function, i.e. the function can either successfully terminate the transaction or return (as in exit to continue control flow after the call at the call site), the optimizer may incorrectly remove state changes before a call to the function.
+
+This happens in cases, in which the state change before the function could be considered redundant, if the fact that the function can terminate the transaction was ignored.
+
+When using the legacy code generator, the Yul optimizer step responsible for the bug will be run on each inline assembly block in isolation and hence the above conditions have to occur inside one inline assembly block. While still possible, this makes the occurrence of the bug with legacy code generation less likely (see the section of Technical Details below for examples of affected assembly snippets).
+
+However, the new via-IR code generation translates the entire contract to Yul and runs the optimizer on all of it. In this case, the bug can be triggered more easily. The following is a minimal example:
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1; // This write used to be removed by the yul optimizer.
+		g();
+		x = 2;
+	}
+	function g() internal {
+		// The important aspect of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which it terminates the transaction, since the state change above will be missing.
+		if (msg.data.length > 4) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result ``f()`` without additional calldata to incorrectly return without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic store removal step, e.g. the following is almost
+identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1;
+		g();
+		x = 2;
+	}
+	function g() internal {
+		if (msg.data.length <= 4)
+			assembly { return(0,0) }
+	}
+}
+```
+
+
+## Technical Details
```

Comment:
I think it would be best to have an "Impact" section that touches upon actual use cases that might be affected and what's the worst thing that can happen to your contract if it's affected.

Reply:
You could say the same thing about many bugs. I think that trying to find a good, practical example actually helps with assessing severity.

But, well, it was pretty hard to come up with something sensible for the previous bug so maybe we can do without it...